[WIP] Make complex->finite in the codebase.

ShubhamKJha opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
**[Do not merge this PR]**
This PR attempts to make `is_complex` imply `is_finite`. The assumption rules used are:
```py
_assume_rules = FactRules([

    'integer        ->  rational',
    'rational       ->  real',
    'rational       ->  algebraic',
    'algebraic      ->  complex',
    'transcendental ==  complex & !algebraic',
    'real           ->  hermitian',
    'imaginary      ->  complex',
    'imaginary      ->  antihermitian',
    'extended_real  ->  commutative',
    'complex        ->  commutative',
    'complex        ->  finite',

    'odd            ==  integer & !even',
    'even           ==  integer & !odd',

    'real           ->  complex',
    'extended_real  ->  real | infinite',
    'real           ==  extended_real & finite',

    'extended_real        ==  extended_negative | zero | extended_positive',
    'extended_negative    ==  extended_nonpositive & extended_nonzero',
    'extended_positive    ==  extended_nonnegative & extended_nonzero',

    'extended_nonpositive ==  extended_real & !extended_positive',
    'extended_nonnegative ==  extended_real & !extended_negative',

    'real           ==  negative | zero | positive',
    'negative       ==  nonpositive & nonzero',
    'positive       ==  nonnegative & nonzero',

    'nonpositive    ==  real & !positive',
    'nonnegative    ==  real & !negative',

    'positive       ==  extended_positive & finite',
    'negative       ==  extended_negative & finite',
    'nonpositive    ==  extended_nonpositive & finite',
    'nonnegative    ==  extended_nonnegative & finite',
    'nonzero        ==  extended_nonzero & finite',

    'zero           ->  even & finite',
    'zero           ==  extended_nonnegative & extended_nonpositive',
    'zero           ==  nonnegative & nonpositive',
    'nonzero        ->  real',

    'prime          ->  integer & positive',
    'composite      ->  integer & positive & !prime',
    '!composite     ->  !positive | !even | prime',

    'irrational     ==  real & !rational',

    'imaginary      ->  !extended_real',

    'infinite       ->  !finite',
    'noninteger     ==  extended_real & !integer',
    'extended_nonzero == extended_real & !zero',
])
```

#### Other comments
This breaks `zoo.is_complex` throughout the codebase. I have tried to do this without including any extra facts (i.e. `extended_complex` for entities like `zoo`)

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
  * The assumptions system is changed so now only finite values will be considered complex. With this `zoo` is no longer complex in this sense. And hence, `zoo.is_complex` now returns `False`. Also, any symbol with assumption `complex = True` would become `finite ` and with `finite = False` will become `complex = False` by default.
  * A number of instances of `_eval_is_complex` is changed to better accommodate the finite nature of them. 
<!-- END RELEASE NOTES -->

ShubhamKJha committed:
changed instances of is_complex with is_complex or is_infinite

ShubhamKJha committed:
Further changes in the codebase to check test failures

ShubhamKJha committed:
Further test fixes to make them run

jksuom commented on sympy/algebras/quaternion.py:
```diff
@@ -247,7 +247,7 @@ def add(self, other):
         if not isinstance(q2, Quaternion):
             if q1.real_field:
-                if q2.is_complex:
+                if q2.is_complex or q2.is_infinite:
```
I don't think that infinite components should be accepted for quaternions. So maybe this could be left as it is.

ShubhamKJha replied:
It seems this has been accepting infinite components till now. Should this be changed back?

jksuom replied:
It has been accepting infinite components but that is probably not intentional. Anyway, I think that only finite components should be allowed in algebras.

asmeurer replied:
Probably quite a few places are like this. For the places where you changed `is_complex` to `is_complex or is_infinite` did you check if it makes sense for infinite quantities?

oscarbenjamin replied:
I think that this should be left as `is_complex`

ShubhamKJha committed:
Fixes failing tests

oscarbenjamin commented on sympy/core/numbers.py:
```diff
@@ -3342,7 +3342,7 @@ class ComplexInfinity(with_metaclass(Singleton, AtomicExpr)):
     is_number = True
     is_prime = False
-    is_complex = True
+    # is_complex = True
```
This should be changed to `= False`

oscarbenjamin commented on sympy/stats/tests/test_continuous_rv.py:
```diff
@@ -42,7 +42,7 @@ def test_single_normal():
     pdf = density(Y)
     x = Symbol('x')
-    assert (pdf(x) ==
+    assert (pdf(x) == \
```
Line continuation isn't needed with brackets. Actually brackets are preferred to line continuation in PEP8

oscarbenjamin commented on sympy/stats/tests/test_continuous_rv.py:
```diff
@@ -31,7 +31,7 @@
 
-
+@XFAIL
 def test_single_normal():
```
Why is this XFAIL?

ShubhamKJha replied:
This test is failing. I am not familiar with the code in the `stats` part, so I couldn't figure it out.
The result I am getting is:
```py
>>> from sympy import *
>>> from sympy.stats import Normal, quantile
>>> x = Symbol('x')
>>> mu = Symbol('mu', real=True)
>>> sigma = Symbol('sigma', positive=True)
>>> X = Normal('x', 0, 1)
>>> Y = X*sigma + mu
>>> quantile(Y)(x)
{sqrt(2)*sigma*(sqrt(2)*mu/(2*sigma) + erfinv(2*x - 1))}

## or
⎧     ⎛√2⋅μ                  ⎞⎫
⎨√2⋅σ⋅⎜──── + erfinv(2⋅x - 1)⎟⎬
⎩     ⎝2⋅σ                   ⎠⎭

```
while the original result is:
```py
>>> Intersection(S.Reals, FiniteSet(sqrt(2)*sigma*(sqrt(2)*mu/(2*sigma) + erfinv(
2*x - 1))))
    ⎧     ⎛√2⋅μ                  ⎞⎫
ℝ ∩ ⎨√2⋅σ⋅⎜──── + erfinv(2⋅x - 1)⎟⎬
    ⎩     ⎝2⋅σ                   ⎠⎭
```

oscarbenjamin replied:
We need to check the assumptions on that expression. The intersection will evaluate if the expression is deemed to be real (or not).

ShubhamKJha replied:
The intersection doesn't evaluate, the operation `quantile(Y)(x)` is resulting the evaluated result only.

czgdp1807 replied:
the result you are getting is correct, may be the code of `quantile` needs some changes in assumptions of variables, try some debugging and check where the `Intersection` is generated.

czgdp1807 replied:
I would say your result looks more intuitive, change the test if needed.

oscarbenjamin replied:
I'm unresolving this because the XFAIL is still there.

ShubhamKJha replied:
XFAIL has been removed

oscarbenjamin commented:
This generally looks good although I left comments

There are lots of places where you've changed e.g. real->extended_real or positive->extended_positive. Are those needed because of the changes here or is that just unfinished business from #16666?

It looks like there are a number of places where you are testing `if x.is_complex or x.is_infinite`. Potentially those would benefit from being `x.is_extended_complex`. How many places are there in the code where `x.is_complex` is used and how many need to also check for `x.is_finite`? I count 40 occurrences of is_complex:
```
$ git grep is_complex | grep -v test_ | wc -l
      40
```

ShubhamKJha commented:
> There are lots of places where you've changed e.g. real->extended_real or positive->extended_positive. Are those needed because of the changes here or is that just unfinished business from #16666?

I tried not to revert any from #16666, most of them appeared to be left unchanged during #16666.

Apart from the occurrences of `is_complex` I also checked the occurrences of `complex` and changed when seemed appropriate. 
The present notion of `complex` is that of `extended_complex`, therefore, I changed the instances of `is_complex` to `is_complex or is_infinite` to better account for `oo`, `-oo`, `zoo`.

ShubhamKJha commented:
Thanks @oscarbenjamin for this review, I will go through the diff for the instances where `is_complex` is actually asking for finite complex.

ShubhamKJha committed:
Updated code as suggested

oscarbenjamin commented:
This branch has merge conflicts now

ShubhamKJha committed:
Merge branch 'master' into complex_finite

ShubhamKJha committed:
merge master into complex_finite

ShubhamKJha committed:
Merge branch 'master' into complex_finite

ShubhamKJha committed:
Update for failing tests

oscarbenjamin commented on sympy/core/assumptions.py:
```diff
@@ -376,7 +376,7 @@ def __init__(cls, *args, **kws):
                 setattr(cls, pname, make_property(fact))
 
-        # Finally, add any missing automagic property (e.g. for Basic)
+        # Finally, add any missing automatic property (e.g. for Basic)
```
I think this is probably deliberately spelled as automagic (which is a word)

oscarbenjamin commented on sympy/core/expr.py:
```diff
@@ -332,7 +332,7 @@ def __ge__(self, other):
             raise TypeError("Invalid comparison %s >= %s" % (self, other))
         for me in (self, other):
-            if me.is_complex and me.is_extended_real is False:
+            if (me.is_complex or me.is_infinite) and me.is_extended_real is False:
```
In general this replacement will not be equivalent since if `x` is complex then `(1/x)` will be complex or infinite but `(1/x).is_complex` and `(1/x).is_infinite` will both return None. It isn't possible to make something equivalent without having extended_complex.

oscarbenjamin replied:
Do we need the `is_complex` check here at all? Shouldn't we just raise if extended_real is False?

oscarbenjamin replied:
(Same goes for the other comparison methods below.)

ShubhamKJha committed:
Added changes as suggested

ShubhamKJha committed:
Merge branch 'master' into complex_finite

ShubhamKJha commented:
ping @oscarbenjamin @asmeurer @jksuom. Are we still in support of this? I think it can be merged.

ShubhamKJha committed:
Merge branch 'complex_finite' of https://github.com/ShubhamKJha/sympy into complex_finite

asmeurer commented:
Let's decide either way on this before the next release.

jksuom commented:
Having complex => finite looks good to me. I think that it should be accepted if possible without too many issues.

asmeurer commented:
There's still a comment on the diff about quaternions not accepting infinite values. Were all the instances where `is_infinite` was added checked to see if infinite really makes sense there? I know it's the same behavior as before, but it would be really confusing to have that explicitly in the code where it doesn't really belong.

oscarbenjamin commented:
I will go through the diff soon but in general I think that this should be merged for the next release. The real->Finite changes are since the last release so merging this as well means they both change simultaneously.

oscarbenjamin commented on sympy/core/tests/test_arit.py:
```diff
@@ -1898,38 +1898,38 @@ def test_mul_coeff():
+    nz = Dummy(real=True, zero=False)
+    r = Dummy(extended_real=True)
+    c = Dummy(extended_real=False, complex=True)
+    c2 = Dummy(extended_real=False, complex=True)
```
Why have these become extended_real rather than real. If complex implies finite then e.g. `extended_real=False,complex=True` should be the same as `real=False,complex=True` since the difference between real and extended_real is the infinities that are excluded by complex=True under this PR.

oscarbenjamin commented on sympy/core/tests/test_assumptions.py:
```diff
@@ -1134,8 +1134,8 @@ def test_issue_16313():
     # with work on PR 16603 it may be changed in future to complex -> finite
-    x = Symbol('x', complex=True, finite=False)
-    y = Symbol('x', real=True, infinite=False)
+    #x = Symbol('x', complex=True, finite=False)
```
This and the comment above can be removed. This whole test function is probably not needed any more although some of the asserts could be moved elsewhere.

oscarbenjamin commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -1219,9 +1230,18 @@ def _eval_is_real(self):
+        arg = self.args[0]
+
+        if arg.is_real and (arg / pi - S.Half).is_integer is False:
+            return True
```
If we had extended_complex then methods like this wouldn't be needed. We could just say that extended_complex is True and let `_eval_is_finite` imply complex.

oscarbenjamin commented:
I guess the only question here is whether we want extended_complex. I think this diff would look very different if we went down that road.

jksuom commented:
I think that symbols are expected to be extended complex by default (though that may not have been explicitly told anywhere). If that is understood, then there will probably be no need for extended_complex assumption.

oscarbenjamin commented:
> I think that symbols are expected to be extended complex by default (though that may not have been explicitly told anywhere). If that is understood, then there will probably be no need for extended_complex assumption.

The question is how do you test for it. The diff here has a load of `if x.is_complex or x.is_finite` but that won't work for say `1/x` where `x` is complex since both conditions will give `None`. Likewise what is the test that differentiates between a vanilla symbol with implicit extended complex assumptions and, say a MatrixSymbol or a quaternion or something?

asmeurer commented:
By default they are commutative, but I guess you can basically see commutative as equal to extended complex. I think that's not actually strictly true because you might have some other commutative objects that aren't extended complex (like some elements of some commutative algebra perhaps). 

I think in practice it means we have to check only for negative conditions, so instead of 

```py
if x.is_complex is True: good
```

use 

```py
if x.is_complex is False: bad
```

That's similar to how we handle functions with implicit domains, for instance, if a function only makes sense for integer inputs, we fail for `is_integer=False` but not for `is_integer=None`.

oscarbenjamin commented:
I think we should merge this. We can add extended complex later if needed. Let's get this in to the next release and let users see it.

oscarbenjamin commented:
Another example:
```julia
In [11]: f = Lambda(x, x**2)                                                                                                      

In [12]: f.is_complex == None                                                                                                     
Out[12]: True

In [13]: exp(f)                                                                                                                   
Out[13]: 
      2
 x ↦ x 
ℯ  
```
What test should exp use to check that its argument is a legit "number"?

Here's another:
```julia
In [14]: exp(sin)                                                                                                                 
Out[14]: 
 sin
ℯ   

In [15]: sin.is_complex                                                                                                           
Out[15]: <property at 0x104f61e08>

In [16]: bool(sin.is_complex)                                                                                                     
Out[16]: True

In [19]: bool(sin.is_commutative)                                                                                                 
Out[19]: True
```

jksuom commented:
We should be able to set `is_commutative` and `is_complex` false for functions. Maybe by defining `_explicit_class_assumptions`.

oscarbenjamin commented:
I still think there's a need to have a property that corresponds to what would be expected as input by most mathematical functions which is basically complex, oo, -oo and zoo and quantities derived from them like `1+I*oo`. The obvious property name would be something like `is_number` but we already have that and it means something different.

ShubhamKJha committed:
Updated according to suggestions

ShubhamKJha committed:
code quality issue

oscarbenjamin commented:
This needs reviving. We shouldn't forget this before the next release. There are some test failures but they probably aren't hard to fix.

The discussion above seems to have gotten bogged down in the question of whether we need extended_complex but I think we can get this merged and worry about whether we need that later.

asmeurer commented:
Yes, this is blocking a release.

ShubhamKJha committed:
Invalid comparisons error should occur only when any of the operands is a complex value
